torch/csrc/cuda/Module.cpp

```diff
@@ -917,6 +920,29 @@ static void registerCudaDeviceProperties(PyObject* module) {
       });
 }
 
+void no_op_delete(void* ptr){};
+
+// We choose to ignore certain blocks that are currently allocated
+// when we set the pool to its checkpoint. For those blocks, we need
+// to swap out the deleter function of their corresponding blocks
+// so that a deallocation is not triggered when they die.
+void swapLiveStorageDeleterFns(
+    const std::vector<c10::StorageImpl*>& stale_live_storages,
+    std::unordered_set<void*> definitely_stale_pointers) {
+  for (c10::StorageImpl* stale_storage : stale_live_storages) {
+    auto ptr = stale_storage->data_ptr().get();
+    auto allocated_pointer = definitely_stale_pointers.find(ptr);
+    TORCH_CHECK(allocated_pointer != definitely_stale_pointers.end());
+    auto t = c10::cuda::CUDACachingAllocator::get();
+    bool succeeded = stale_storage->data_ptr().compare_exchange_deleter(
+        t->raw_deleter(), &no_op_delete);
+
+    TORCH_CHECK(
```

Comment:
So, when can this happen? Is this if I pass garbage tensors to the checkpoint function? If I only ever pass tensors that are allocated out of the pool, the deleter should always be as expected, right? Can we describe the preconditions for the checkpoint tensor list?

Reply:
This error means that you had a non-standard deleter fn, and so it's not safe for us to blindly remove it. As the comment in Allocator.h says:

```
   * Compare the deleter in a DataPtr to expected_deleter.
   * If it matches, replace the deleter with new_deleter
   * and return true; otherwise, does nothing and returns
   * false.
   *
   * In general, it is not safe to unconditionally set the
   * deleter on a DataPtr, because you don't know what
   * the deleter is, and thus will have a hard time properly
   * disposing of the deleter without storing the original
   * deleter
   ```